pkg/executor/aggregate/agg_spill_test.go

```diff
@@ -0,0 +1,421 @@
+	ctx.GetSessionVars().MaxChunkSize = chkSize
+	ctx.GetSessionVars().MemTracker = memory.NewTracker(memory.LabelForSession, hardLimitBytesNum)
+	ctx.GetSessionVars().TrackAggregateMemoryUsage = true
+	ctx.GetSessionVars().EnableParallelHashaggSpill = true
+	ctx.GetSessionVars().StmtCtx.MemTracker = memory.NewTracker(memory.LabelForSQLText, -1)
+	ctx.GetSessionVars().StmtCtx.MemTracker.AttachTo(ctx.GetSessionVars().MemTracker)
+	ctx.GetSessionVars().MemTracker.SetActionOnExceed(newRootExceedAction)
+}
+
+func executeCorrecResultTest(t *testing.T, ctx *mock.Context, dataSource *testutil.MockDataSource, result map[string]float64) {
+	aggExec := buildHashAggExecutor(t, ctx, dataSource)
+	dataSource.PrepareChunks()
+	tmpCtx := context.Background()
+	chk := exec.NewFirstChunk(aggExec)
+	resContainer := resultsContainer{}
+	aggExec.Open(tmpCtx)
+
+	for {
+		err := aggExec.Next(tmpCtx, chk)
+		require.Equal(t, nil, err)
+		if chk.NumRows() == 0 {
+			break
+		}
+		resContainer.add(chk)
+		chk.Reset()
+	}
+	aggExec.Close()
+
+	require.True(t, aggExec.IsSpillTriggeredForTest())
+	require.True(t, resContainer.check(result))
+}
+
+func fallBackActionTest(t *testing.T) {
+	newRootExceedAction := new(testutil.MockActionOnExceed)
+	hardLimitBytesNum := int64(6000000)
+
+	ctx := mock.NewContext()
+	initCtx(ctx, newRootExceedAction, hardLimitBytesNum, 4096)
+
+	// Consume lots of memory in advance to help to trigger fallback action.
+	ctx.GetSessionVars().MemTracker.Consume(int64(float64(hardLimitBytesNum) * 0.799999))
+
+	rowNum := 10000 + rand.Intn(10000)
+	ndv := 5000 + rand.Intn(5000)
+	col1, col2 := generateData(rowNum, ndv)
+	opt := getMockDataSourceParameters(ctx)
+	dataSource := buildMockDataSource(opt, col1, col2)
+
+	aggExec := buildHashAggExecutor(t, ctx, dataSource)
+	dataSource.PrepareChunks()
+	tmpCtx := context.Background()
+	chk := exec.NewFirstChunk(aggExec)
+	resContainer := resultsContainer{}
+	aggExec.Open(tmpCtx)
+
+	for {
+		aggExec.Next(tmpCtx, chk)
+		if chk.NumRows() == 0 {
+			break
+		}
+		resContainer.add(chk)
+		chk.Reset()
+	}
+	aggExec.Close()
+	require.Less(t, 0, newRootExceedAction.GetTriggeredNum())
+}
+
+func randomFailTest(t *testing.T, ctx *mock.Context, dataSource *testutil.MockDataSource) {
+	dataSource.PrepareChunks()
+	aggExec := buildHashAggExecutor(t, ctx, dataSource)
+	tmpCtx := context.Background()
+	chk := exec.NewFirstChunk(aggExec)
+	aggExec.Open(tmpCtx)
+
+	goRoutineWaiter := sync.WaitGroup{}
+	goRoutineWaiter.Add(1)
+	defer goRoutineWaiter.Wait()
+
+	once := sync.Once{}
+
+	go func() {
+		time.Sleep(time.Duration(rand.Int31n(300)) * time.Millisecond)
+		once.Do(func() {
+			aggExec.Close()
+		})
+		goRoutineWaiter.Done()
+	}()
+
+	for {
+		err := aggExec.Next(tmpCtx, chk)
+		if err != nil {
+			once.Do(func() {
+				err = aggExec.Close()
+				require.Equal(t, nil, err)
+			})
+			break
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+		chk.Reset()
+	}
+	once.Do(func() {
+		aggExec.Close()
+	})
```

Comment:
Why there are 3 aggExec.Close?

Reply:
> Why there are 3 aggExec.Close?

The first `Close()` mock the close called by somewhere else when agg is running.

The second close the executor when agg encounters errors.

The third is the normal `Close()`